### PR TITLE
Deprecate async def queries and disallow most workflow operations in read-only context

### LINKED ISSUE #250: [Feature Request] Warn/error on invalid query use
### Describe the solution you'd like

* If a query is an `async def` we should warn by default but allow overriding/silencing that warning
* If a query uses workflow constructs like `wait_condition` we should error
* If core is not already doing it for us, if a query does anything to create commands we should error

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

* Deprecated `async def` queries
* Raise a read-only-context error for doing anything invalid in a query (and it's also ready for update validators when we get there)

## Checklist

1. Closes #250

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -566,7 +566,8 @@ Here are the decorators that can be applied:
   * Return value is ignored
 * `@workflow.query` - Defines a method as a query
   * All the same constraints as `@workflow.signal` but should return a value
-  * Temporal queries should never mutate anything in the workflow
+  * Should not be `async`
```
Comment (Quinn-With-Two-Ns) [inline README.md:569]: This is a breaking change correct?

File: temporalio/worker/_workflow_instance.py [hunk 2]
```diff
@@ -1153,8 +1171,24 @@ async def run_child() -> Any:
...
+    def _as_read_only(self) -> Iterator[None]:
+        prev_val = self._read_only
+        self._read_only = True
+        try:
+            yield None
+        finally:
+            self._read_only = prev_val
+
+    def _assert_not_read_only(self, action_attempted: str) -> None:
+        if self._read_only:
```
[the remark this answers is not in the document]
Comment (Quinn-With-Two-Ns) [reply temporalio/worker/_workflow_instance.py:1187]: oh ignore I missed where you dealt with random

File: temporalio/worker/_workflow_instance.py [hunk 3]
```diff
@@ -1258,6 +1292,7 @@ def _register_task(
         *,
         name: Optional[str],
     ) -> None:
+        self._assert_not_read_only("create task")
```
Comment (Quinn-With-Two-Ns) [inline temporalio/worker/_workflow_instance.py:1295]: nit: "cancel workflow"?

File: README.md (hunk 1 above)
Comment (cretz) [reply README.md:569]: I just deprecate and send a warning

File: temporalio/worker/_workflow_instance.py (hunk 3 above)
Comment (Quinn-With-Two-Ns) [reply temporalio/worker/_workflow_instance.py:1295]: oh that's my bad, github UI made this look like the check for `_cancel_external_workflow`

**Review by Quinn-With-Two-Ns — APPROVED** [review]
nice, easier to do with no side effects

**Pull request merged**